File: packages/server/src/services/prompt-service.ts
import { promptStorage, type PromptsStorage, type PromptProjectsStorage } from '@/utils/storage/prompt-storage'
import {
  CreatePromptBody,
  UpdatePromptBody,
  Prompt,
  PromptSchema,
  PromptProject,
  PromptProjectSchema
} from 'shared/src/schemas/prompt.schemas'

import { normalizeToUnixMs } from '@/utils/parse-timestamp'
import { ApiError } from 'shared'
import { ZodError } from 'zod'

export async function createPrompt(data: CreatePromptBody): Promise<Prompt> {
  const promptId = promptStorage.generateId()
  const now = normalizeToUnixMs(new Date())

  const newPromptData: Prompt = {
    id: promptId,
    name: data.name,
    content: data.content,
    created: now,
    updated: now
    // projectId is not part of the core Prompt schema, handled by associations
  }

  try {
    PromptSchema.parse(newPromptData) // Validate before adding to storage
  } catch (error) {
    if (error instanceof ZodError) {
      console.error(`Validation failed for new prompt data: ${error.message}`, error.flatten().fieldErrors)
      throw new ApiError(500, `Internal validation error creating prompt.`, 'PROMPT_VALIDATION_ERROR', error.flatten().fieldErrors)
    }
    throw error // Should not happen if data is constructed correctly
  }

  const allPrompts = await promptStorage.readPrompts()

  if (allPrompts[promptId]) {
    throw new ApiError(500, `Prompt ID conflict for ${promptId}`, 'PROMPT_ID_CONFLICT')
  }

  allPrompts[promptId] = newPromptData
  await promptStorage.writePrompts(allPrompts)

  if (data.projectId) {
    await addPromptToProject(newPromptData.id, data.projectId)
  }

  return newPromptData
}

export async function addPromptToProject(promptId: number, projectId: number): Promise<void> {
  const allPrompts = await promptStorage.readPrompts()
  if (!allPrompts[promptId]) {
    throw new ApiError(404, `Prompt with ID ${promptId} not found.`, 'PROMPT_NOT_FOUND')
  }

  let promptProjects = await promptStorage.readPromptProjects()

  // Check if association already exists
  const existingLink = promptProjects.find(link => link.promptId === promptId && link.projectId === projectId)
  if (existingLink) {
    return // Association already exists, do nothing
  }

  // In the previous DB version, this was effectively an "set project for prompt"
  // For a file-based many-to-many, we usually add.
  // The old code `db.prepare('DELETE FROM prompt_projects WHERE prompt_id = ?').run(promptId)`
  // would remove the prompt from ALL projects before adding to one.
  // This is a change in behavior if we simply add.
  // Let's stick to the previous behavior for now: a prompt is associated with one project via this call.
  // If multiple projects are desired, a different method or logic modification is needed.
  // The schema `CreatePromptBodySchema` implies one project association at creation.
  // The `listPromptsByProject` and `removePromptFromProject` suggest many-to-many is possible.

  // Re-evaluating: The original DB `addPromptToProject` did NOT remove all other associations.
  // It only added a new one. The `DELETE FROM prompt_projects WHERE prompt_id = ?` was in `createPrompt` if a `projectId` was given.
  // Let's look at the `createPrompt`'s use of `addPromptToProject`.
  // The DB `addPromptToProject`:
  // `db.prepare('DELETE FROM prompt_projects WHERE prompt_id = ?').run(promptId)` THIS IS THE LINE.
  // So yes, it first removes all project associations for that prompt, then adds the new one.
  // This means a prompt is assigned to at most one project via the `createPrompt` or direct `addPromptToProject` flow.

  // To replicate: filter out existing links for this promptId, then add the new one.
  promptProjects = promptProjects.filter(link => link.promptId !== promptId)

  const newLink: PromptProject = {
    id: promptStorage.generateId(), // ID for the association itself
    promptId: promptId,
    projectId: projectId
  }

  try {
    PromptProjectSchema.parse(newLink)
  } catch (error) {
    if (error instanceof ZodError) {
      console.error(`Validation failed for new prompt-project link: ${error.message}`, error.flatten().fieldErrors)
      throw new ApiError(500, `Internal validation error linking prompt to project.`, 'PROMPT_LINK_VALIDATION_ERROR', error.flatten().fieldErrors)
    }
    throw error
  }

  promptProjects.push(newLink)
  await promptStorage.writePromptProjects(promptProjects)
}

export async function removePromptFromProject(promptId: number, projectId: number): Promise<void> {
  let promptProjects = await promptStorage.readPromptProjects()
  const initialLinkCount = promptProjects.length

  promptProjects = promptProjects.filter(link => !(link.promptId === promptId && link.projectId === projectId))

  if (promptProjects.length === initialLinkCount) {
    const allPrompts = await promptStorage.readPrompts()
    if (!allPrompts[promptId]) {
      throw new ApiError(404, `Prompt with ID ${promptId} not found.`, 'PROMPT_NOT_FOUND')
    }
    // If prompt exists, but link didn't, then the association was not found.
    throw new ApiError(
      404,
      `Association between prompt ${promptId} and project ${projectId} not found.`,
      'PROMPT_PROJECT_LINK_NOT_FOUND'
    )
  }

  await promptStorage.writePromptProjects(promptProjects)
}

export async function getPromptById(promptId: number): Promise<Prompt> {
  const allPrompts = await promptStorage.readPrompts()
  const found = allPrompts[promptId]
  if (!found) {
    throw new ApiError(404, `Prompt with ID ${promptId} not found.`, 'PROMPT_NOT_FOUND')
  }
  return found
}

export async function listAllPrompts(): Promise<Prompt[]> {
  const allPromptsData = await promptStorage.readPrompts()
  const promptList = Object.values(allPromptsData)
  // Optional: sort if needed, e.g., by upated or name
  promptList.sort((a, b) => (a.name || '').localeCompare(b.name || ''))
  return promptList
}

export const getPromptsByIds = async (promptIds: string[]): Promise<Prompt[]> => {
  const allPrompts = await promptStorage.readPrompts()
  return promptIds.map(id => allPrompts[id]).filter(Boolean) as Prompt[]
}

export async function listPromptsByProject(projectId: number): Promise<Prompt[]> {
  const promptProjects = await promptStorage.readPromptProjects()
  const relevantPromptIds = promptProjects
    .filter(link => link.projectId === projectId)
    .map(link => link.promptId)

  if (relevantPromptIds.length === 0) {
    return []
  }
  const allPrompts = await promptStorage.readPrompts()
  return relevantPromptIds.map(promptId => allPrompts[promptId]).filter(Boolean) as Prompt[] // filter(Boolean) to remove undefined if a link exists for a deleted prompt
}

export async function updatePrompt(promptId: number, data: UpdatePromptBody): Promise<Prompt> {
  const allPrompts = await promptStorage.readPrompts()
  const existingPrompt = allPrompts[promptId]

  if (!existingPrompt) {
    throw new ApiError(404, `Prompt with ID ${promptId} not found for update.`, 'PROMPT_NOT_FOUND')
  }

  const updatedPromptData: Prompt = {
    ...existingPrompt,
    name: data.name ?? existingPrompt.name,
    content: data.content ?? existingPrompt.content,
    updated: normalizeToUnixMs(new Date())
  }

  try {
    PromptSchema.parse(updatedPromptData)
  } catch (error) {
    if (error instanceof ZodError) {
      console.error(`Validation failed updating prompt ${promptId}: ${error.message}`, error.flatten().fieldErrors)
      throw new ApiError(500, `Internal validation error updating prompt.`, 'PROMPT_VALIDATION_ERROR', error.flatten().fieldErrors)
    }
    throw error
  }

  allPrompts[promptId] = updatedPromptData
  await promptStorage.writePrompts(allPrompts)
  return updatedPromptData
}

export async function deletePrompt(promptId: number): Promise<boolean> {
  const allPrompts = await promptStorage.readPrompts()
  if (!allPrompts[promptId]) {
    return false // Prompt not found, nothing to delete
  }

  delete allPrompts[promptId]
  await promptStorage.writePrompts(allPrompts)

  // Also remove any associations for this prompt
  let promptProjects = await promptStorage.readPromptProjects()
  const initialLinkCount = promptProjects.length
  promptProjects = promptProjects.filter(link => link.promptId !== promptId)

  if (promptProjects.length < initialLinkCount) {
    await promptStorage.writePromptProjects(promptProjects)
  }

  return true
}

export async function getPromptProjects(promptId: number): Promise<PromptProject[]> {
  const promptProjects = await promptStorage.readPromptProjects()
  return promptProjects.filter(link => link.promptId === promptId)
}

File: packages/python_backend/app/services/prompt_service.py
from datetime import datetime, timezone
from typing import List, Optional
from fastapi import HTTPException
from pydantic import ValidationError

from app.schemas.prompt_schemas import (
    Prompt,
    CreatePromptBody,
    UpdatePromptBody,
    PromptProject
)
from app.utils.storage.prompt_storage import prompt_storage_util, PromptsStorage, PromptProjectsStorage
from app.utils.storage_timestap_utils import convert_timestamp_to_ms_int

# Last 5 changes (from .ts):
# 1. Converted createPrompt
# 2. Converted addPromptToProject and its specific logic for single project association
# 3. Converted list, get, update, delete operations for prompts
# 4. Converted project-specific prompt operations (listByProject, removeFromProject)
# 5. Adapted error handling to FastAPI's HTTPException
# 6. Changed ID and timestamp generation/handling to use integer Unix ms.

async def create_prompt(data: CreatePromptBody) -> Prompt:
    prompt_id = prompt_storage_util.generate_id()
    now_ms = prompt_storage_util.generate_id()

    new_prompt_data_dict = {
        "id": prompt_id,
        "name": data.name,
        "content": data.content,
        "created": now_ms,
        "updated": now_ms
        # projectId is handled by association, not part of core Prompt model attribute in Python version
    }

    try:
        # Validate with Pydantic model before adding to storage
        new_prompt = Prompt.model_validate(new_prompt_data_dict)
    except ValidationError as e:
        print(f"Validation failed for new prompt data: {e.errors()}")
        raise HTTPException(status_code=500, detail={"message": "Internal validation error creating prompt.", "code": "PROMPT_VALIDATION_ERROR", "errors": e.errors()})

    all_prompts = await prompt_storage_util.read_prompts()

    if prompt_id in all_prompts:
        raise HTTPException(status_code=500, detail={"message": f"Prompt ID conflict for {prompt_id}", "code": "PROMPT_ID_CONFLICT"})

    all_prompts[prompt_id] = new_prompt
    await prompt_storage_util.write_prompts(all_prompts)

    if data.project_id:
        project_id_int = int(data.project_id) if isinstance(data.project_id, str) else data.project_id
        await add_prompt_to_project(new_prompt.id, project_id_int)

    return new_prompt

async def add_prompt_to_project(prompt_id: int, project_id: int) -> None:
    all_prompts = await prompt_storage_util.read_prompts()
    if prompt_id not in all_prompts:
        raise HTTPException(status_code=404, detail={"message": f"Prompt with ID {prompt_id} not found.", "code": "PROMPT_NOT_FOUND"})

    prompt_projects = await prompt_storage_util.read_prompt_projects()

    # Replicate original TS logic: a prompt is assigned to at most one project via this flow.
    # Filter out existing links for this promptId, then add the new one.
    updated_prompt_projects = [link for link in prompt_projects if link.prompt_id != prompt_id]

    new_link_data = {
        "id": prompt_storage_util.generate_id(),
        "prompt_id": prompt_id,
        "project_id": project_id,
        "created": prompt_storage_util.generate_id()
    }
    try:
        new_link = PromptProject.model_validate(new_link_data)
    except ValidationError as e:
        print(f"Validation failed for new prompt-project link: {e.errors()}")
        raise HTTPException(status_code=500, detail={"message": "Internal validation error linking prompt to project.", "code": "PROMPT_LINK_VALIDATION_ERROR", "errors": e.errors()})
    
    updated_prompt_projects.append(new_link)
    await prompt_storage_util.write_prompt_projects(updated_prompt_projects)

async def remove_prompt_from_project(prompt_id: int, project_id: int) -> None:
    prompt_projects = await prompt_storage_util.read_prompt_projects()
    initial_link_count = len(prompt_projects)

    updated_prompt_projects = [link for link in prompt_projects if not (link.prompt_id == prompt_id and link.project_id == project_id)]

    if len(updated_prompt_projects) == initial_link_count:
        all_prompts = await prompt_storage_util.read_prompts()
        if prompt_id not in all_prompts:
            raise HTTPException(status_code=404, detail={"message": f"Prompt with ID {prompt_id} not found.", "code": "PROMPT_NOT_FOUND"})
        raise HTTPException(
            status_code=404,
            detail={"message": f"Association between prompt {prompt_id} and project {project_id} not found.", "code": "PROMPT_PROJECT_LINK_NOT_FOUND"}
        )

    await prompt_storage_util.write_prompt_projects(updated_prompt_projects)

async def get_prompt_by_id(prompt_id: int) -> Prompt:
    all_prompts = await prompt_storage_util.read_prompts()
    found = all_prompts.get(prompt_id)
    if not found:
        raise HTTPException(status_code=404, detail={"message": f"Prompt with ID {prompt_id} not found.", "code": "PROMPT_NOT_FOUND"})
    return found # Already a Prompt object due to read_prompts validation

async def list_all_prompts() -> List[Prompt]:
    all_prompts_data = await prompt_storage_util.read_prompts()
    prompt_list = list(all_prompts_data.values())
    prompt_list.sort(key=lambda p: (p.name or "").lower()) # Sort by name, case-insensitive
    return prompt_list

async def get_prompts_by_ids(prompt_ids: List[int]) -> List[Prompt]:
    all_prompts = await prompt_storage_util.read_prompts()
    return [all_prompts[id_] for id_ in prompt_ids if id_ in all_prompts]

async def list_prompts_by_project(project_id: int) -> List[Prompt]:
    prompt_projects = await prompt_storage_util.read_prompt_projects()
    relevant_prompt_ids = [link.prompt_id for link in prompt_projects if link.project_id == project_id]

    if not relevant_prompt_ids:
        return []
    
    all_prompts = await prompt_storage_util.read_prompts()
    # Filter out None if a link exists for a deleted prompt
    return [all_prompts[prompt_id] for prompt_id in relevant_prompt_ids if prompt_id in all_prompts]

async def update_prompt(prompt_id: int, data: UpdatePromptBody) -> Prompt:
    all_prompts = await prompt_storage_util.read_prompts()
    existing_prompt = all_prompts.get(prompt_id)

    if not existing_prompt:
        raise HTTPException(status_code=404, detail={"message": f"Prompt with ID {prompt_id} not found for update.", "code": "PROMPT_NOT_FOUND"})

    update_data_dict = data.model_dump(exclude_unset=True) # Get only provided fields
    
    updated_prompt_data = existing_prompt.model_copy(update=update_data_dict)
    updated_prompt_data.updated = prompt_storage_util.generate_id()

    try:
        # Re-validate the whole model. Pydantic will ensure type correctness.
        # UpdatePromptBody already has a validator for at least one field present.
        Prompt.model_validate(updated_prompt_data.model_dump()) 
    except ValidationError as e:
        print(f"Validation failed updating prompt {prompt_id}: {e.errors()}")
        raise HTTPException(status_code=500, detail={"message": "Internal validation error updating prompt.", "code": "PROMPT_VALIDATION_ERROR", "errors": e.errors()})

    all_prompts[prompt_id] = updated_prompt_data
    await prompt_storage_util.write_prompts(all_prompts)
    return updated_prompt_data

async def delete_prompt(prompt_id: int) -> bool:
    all_prompts = await prompt_storage_util.read_prompts()
    if prompt_id not in all_prompts:
        return False # Prompt not found, nothing to delete

    del all_prompts[prompt_id]
    await prompt_storage_util.write_prompts(all_prompts)

    # Also remove any associations for this prompt
    prompt_projects = await prompt_storage_util.read_prompt_projects()
    initial_link_count = len(prompt_projects)
    updated_prompt_projects = [link for link in prompt_projects if link.prompt_id != prompt_id]

    if len(updated_prompt_projects) < initial_link_count:
        await prompt_storage_util.write_prompt_projects(updated_prompt_projects)

    return True

async def get_prompt_projects(prompt_id: int) -> List[PromptProject]:
    prompt_projects = await prompt_storage_util.read_prompt_projects()
    return [link for link in prompt_projects if link.prompt_id == prompt_id]
